File: app/src/main/java/com/arturogutierrez/openticator/view/activity/BaseActivity.java
package com.arturogutierrez.openticator.view.activity;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.arturogutierrez.openticator.R;
import com.arturogutierrez.openticator.application.OpenticatorApplication;
import com.arturogutierrez.openticator.di.component.ApplicationComponent;
import com.arturogutierrez.openticator.di.module.ActivityModule;
import com.arturogutierrez.openticator.domain.navigator.Navigator;
import javax.inject.Inject;

public abstract class BaseActivity extends AppCompatActivity {

  @Bind(R.id.toolbar)
  protected Toolbar toolbar;

  @Inject
  protected Navigator navigator;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setContentView(getLayoutResource());

    getApplicationComponent().inject(this);
    ButterKnife.bind(this);

    configureToolbar();
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == android.R.id.home) {
      finish();
      return true;
    }

    return super.onOptionsItemSelected(item);
  }

  protected ApplicationComponent getApplicationComponent() {
    return ((OpenticatorApplication) getApplication()).getApplicationComponent();
  }

  protected ActivityModule getActivityModule() {
    return new ActivityModule(this);
  }

  protected abstract int getLayoutResource();

  protected void addFragment(int containerViewId, Fragment fragment) {
    FragmentTransaction fragmentTransaction = this.getSupportFragmentManager().beginTransaction();
    fragmentTransaction.add(containerViewId, fragment);
    fragmentTransaction.commit();
  }

  private void configureToolbar() {
    setSupportActionBar(toolbar);
  }
}

File: app/src/main/kotlin/com/arturogutierrez/openticator/view/activity/BaseActivity.kt
package com.arturogutierrez.openticator.view.activity

import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import com.arturogutierrez.openticator.R
import com.arturogutierrez.openticator.application.OpenticatorApplication
import com.arturogutierrez.openticator.di.module.ActivityModule
import com.arturogutierrez.openticator.domain.navigator.Navigator
import com.arturogutierrez.openticator.helpers.consume
import com.arturogutierrez.openticator.helpers.executeTransaction
import org.jetbrains.anko.find
import org.jetbrains.anko.findOptional
import javax.inject.Inject

abstract class BaseActivity : AppCompatActivity() {

  val coordinatorLayout by lazy { find<CoordinatorLayout>(R.id.coordinator_layout) }

  protected val toolbar by lazy { findOptional<Toolbar>(R.id.toolbar) }

  @Inject
  lateinit var navigator: Navigator

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(layoutResource)
    applicationComponent.inject(this)
    configureToolbar()
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    return when (item.itemId) {
      android.R.id.home -> consume { finish() }
      else -> super.onOptionsItemSelected(item)
    }
  }

  protected val applicationComponent by lazy { (application as OpenticatorApplication).applicationComponent }

  protected val activityModule by lazy { ActivityModule(this) }

  protected abstract val layoutResource: Int

  protected fun replaceFragment(containerViewId: Int, fragment: Fragment) {
    supportFragmentManager.executeTransaction {
      replace(containerViewId, fragment)
    }
  }

  private fun configureToolbar() {
    toolbar.let { setSupportActionBar(it) }
  }
}
